Node identity for bench

KirillLykov opened the pull request:
#### Problem

Currently node_id is generated randomly, this complicates testing if we want to use staked connection with bench-tps.
~~I guess it might be possible to get `bind_address` in the code using some wrapper for `getifaddrs`, yet it looks like we usually specify this as cli argument.~~

#### Summary of Changes

* add two additional parameters `bind_address` and `client_node_id` to cli
* add code to get total active stake and stake information from the network
* use this data when constructing `ConnectionCache`

pgarg66 commented on bench-tps/src/cli.rs:
```diff
@@ -355,6 +359,21 @@ pub fn build_args<'a>(version: &'_ str) -> App<'a, '_> {
+        )
+        .arg(
+            Arg::with_name("client_node_id")
+                .long("client-node-id")
```
Can we use "identity" CLI argument instead of adding a new one?

KirillLykov replied:
This `identity` is mislead name, I would rename it to something like `funding-authority` because it is currently used to fund created accounts: https://github.com/solana-labs/solana/blob/master/bench-tps/src/bench.rs#LL996C13-L996C13

But I will have to use different name for the current node_id because if I will use identity and rename identity to something else, this could be misleading for existing users of this argument if any

pgarg66 replied:
Ok, that makes sense.

KirillLykov commented on bench-tps/src/main.rs:
```diff
@@ -52,7 +62,16 @@ fn create_client(
+                        .update_client_certificate(client_node_id, bind_address)
+                        .expect("Failed to update QUIC client certificates");
+
+                    let staked_nodes = Arc::new(RwLock::new(StakedNodes::default()));
```
In the unit tests, I see that both `total_stake` and `pubkey_stake_map` are set.
If this is true, shall I pass this information as cli arguments to bench-tps? Like:
```C++
                    let staked_nodes = Arc::new(RwLock::new(StakedNodes::default()));
                    connection_cache.set_staked_nodes(&staked_nodes, &client_node_id.pubkey());
                    staked_nodes.write().unwrap().total_stake = total_stake; // <- here

                    staked_nodes
                        .write()
                        .unwrap()
                        .pubkey_stake_map
                        .insert(client_node_id.pubkey(), stake); // <- here
```

pgarg66 replied:
Yes, taking these values as CLI argument would be good.

KirillLykov commented:
For myself: don't forget to move `ConnectionCache` outside of the match expression to use these parameters regardless of client type

KirillLykov commented:
@pgarg66 I've added `total_stake` and `stake` to the cli and also propagated them to the `ConnectionCache` setup. 
Yet I'm not sure now if we need to have `stake` and `total_stake` because `StackedNodes` data structure we set is aware only about this particular client and hence `total_stake` should be equal to `stake` (?)

pgarg66 commented:
> @pgarg66 I've added `total_stake` and `stake` to the cli and also propagated them to the `ConnectionCache` setup. Yet I'm not sure now if we need to have `stake` and `total_stake` because `StackedNodes` data structure we set is aware only about this particular client and hence `total_stake` should be equal to `stake` (?)



> @pgarg66 I've added `total_stake` and `stake` to the cli and also propagated them to the `ConnectionCache` setup. Yet I'm not sure now if we need to have `stake` and `total_stake` because `StackedNodes` data structure we set is aware only about this particular client and hence `total_stake` should be equal to `stake` (?)

@KirillLykov, this function is used for computing number of streams a client can open.
https://github.com/solana-labs/solana/blob/c06053f5056b9f5a2bf2c6ff15b8bcf64306e7af/client/src/connection_cache.rs#L143

It uses client's stake, and total stake in the computation. So need both values to have an ideal test condition.
Total stake should be equal to the amount that staked in the cluster. Local stake should be how much is staked on the node whose identity you are using in the client.

KirillLykov commented:
> > @pgarg66 I've added `total_stake` and `stake` to the cli and also propagated them to the `ConnectionCache` setup. Yet I'm not sure now if we need to have `stake` and `total_stake` because `StackedNodes` data structure we set is aware only about this particular client and hence `total_stake` should be equal to `stake` (?)
> 
> > @pgarg66 I've added `total_stake` and `stake` to the cli and also propagated them to the `ConnectionCache` setup. Yet I'm not sure now if we need to have `stake` and `total_stake` because `StackedNodes` data structure we set is aware only about this particular client and hence `total_stake` should be equal to `stake` (?)
> 
> @KirillLykov, this function is used for computing number of streams a client can open.
> 
> https://github.com/solana-labs/solana/blob/c06053f5056b9f5a2bf2c6ff15b8bcf64306e7af/client/src/connection_cache.rs#L143
> 
> It uses client's stake, and total stake in the computation. So need both values to have an ideal test condition. Total stake should be equal to the amount that staked in the cluster. Local stake should be how much is staked on the node whose identity you are using in the client.

In this case, a client can specify value of total_stake which is smaller than the real value of this parameter to increase number of concurrent streams on the client side to max (ie `total_stale == stake`). I suppose that in this case validator will drop extra streams from this client.

I wonder if there is a way to request total stake from network so there is no need in additional argument.

pgarg66 commented:
> > > @pgarg66 I've added `total_stake` and `stake` to the cli and also propagated them to the `ConnectionCache` setup. Yet I'm not sure now if we need to have `stake` and `total_stake` because `StackedNodes` data structure we set is aware only about this particular client and hence `total_stake` should be equal to `stake` (?)
> > 
> > 
> > > @pgarg66 I've added `total_stake` and `stake` to the cli and also propagated them to the `ConnectionCache` setup. Yet I'm not sure now if we need to have `stake` and `total_stake` because `StackedNodes` data structure we set is aware only about this particular client and hence `total_stake` should be equal to `stake` (?)
> > 
> > 
> > @KirillLykov, this function is used for computing number of streams a client can open.
> > https://github.com/solana-labs/solana/blob/c06053f5056b9f5a2bf2c6ff15b8bcf64306e7af/client/src/connection_cache.rs#L143
> > 
> > It uses client's stake, and total stake in the computation. So need both values to have an ideal test condition. Total stake should be equal to the amount that staked in the cluster. Local stake should be how much is staked on the node whose identity you are using in the client.
> 
> In this case, a client can specify value of total_stake which is smaller than the real value of this parameter to increase number of concurrent streams on the client side to max (ie `total_stale == stake`). I suppose that in this case validator will drop extra streams from this client.
> 
> I wonder if there is a way to request total stake from network so there is no need in additional argument.

I think CLI has a few commands that can give you stake information. So client can use that to find the stake (total, as well it's own).

For example, these commands return stake information

```
solana stake-history
solana stakes
```

KirillLykov commented:
> > > > @pgarg66 I've added `total_stake` and `stake` to the cli and also propagated them to the `ConnectionCache` setup. Yet I'm not sure now if we need to have `stake` and `total_stake` because `StackedNodes` data structure we set is aware only about this particular client and hence `total_stake` should be equal to `stake` (?)
> > > 
> > > 
> > > > @pgarg66 I've added `total_stake` and `stake` to the cli and also propagated them to the `ConnectionCache` setup. Yet I'm not sure now if we need to have `stake` and `total_stake` because `StackedNodes` data structure we set is aware only about this particular client and hence `total_stake` should be equal to `stake` (?)
> > > 
> > > 
> > > @KirillLykov, this function is used for computing number of streams a client can open.
> > > https://github.com/solana-labs/solana/blob/c06053f5056b9f5a2bf2c6ff15b8bcf64306e7af/client/src/connection_cache.rs#L143
> > > 
> > > It uses client's stake, and total stake in the computation. So need both values to have an ideal test condition. Total stake should be equal to the amount that staked in the cluster. Local stake should be how much is staked on the node whose identity you are using in the client.
> > 
> > 
> > In this case, a client can specify value of total_stake which is smaller than the real value of this parameter to increase number of concurrent streams on the client side to max (ie `total_stale == stake`). I suppose that in this case validator will drop extra streams from this client.
> > I wonder if there is a way to request total stake from network so there is no need in additional argument.
> 
> I think CLI has a few commands that can give you stake information. So client can use that to find the stake (total, as well it's own).
> 
> For example, these commands return stake information
> 
> ```
> solana stake-history
> solana stakes
> ```

I thought about requesting this information in the code instead of taking it from cli, but probably better to go with these changes as it is and as a follow up add stake request in the code.

CriesofCarrots commented on bench-tps/src/main.rs:
```diff
@@ -19,43 +19,75 @@ use {
+    match use_quic {
+        true => {
+            let staked_nodes = Arc::new(RwLock::new(StakedNodes::default()));
+            staked_nodes.write().unwrap().total_stake = total_stake;
+            staked_nodes
```
Could you just grab a single write lock?

CriesofCarrots commented on bench-tps/src/cli.rs:
```diff
@@ -353,6 +361,39 @@ pub fn build_args<'a>(version: &'_ str) -> App<'a, '_> {
+                .long("client-node-id")
+                .value_name("PATH")
+                .takes_value(true)
+                .help("File containing a node id (keypair) to communicate with validators"),
```
Could you make this help text more descriptive? Why would I need a keypair to communicate with validators?

CriesofCarrots commented on bench-tps/src/cli.rs:
```diff
@@ -353,6 +361,39 @@ pub fn build_args<'a>(version: &'_ str) -> App<'a, '_> {
+                .long("client-node-stake")
+                .value_name("LAMPORTS")
+                .takes_value(true)
+                .help(
+                    "Stake of the client_node_id",
+                ),
+        )
+        .arg(
+            Arg::with_name("client_node_total_stake")
+                .long("client-node-total-stake")
+                .value_name("LAMPORTS")
+                .takes_value(true)
+                .help(
+                    "Total stake of the client_node_id",
```
Can you edit these help texts to explain the difference between these two args?

lijunwangs replied:
Why do we need these two, can we obtain from the network?

KirillLykov replied:
I wanted to add it as follow up PR, but can add here, maybe even simpler.

My general concern is that I think that we don't need to set up CC parameters on client side. Because there is one way quic connections and client doesn't receive much so receive window should not really matter. I don't observe effect of setting up these parameter on the client in my limited experiments (haven't dig into that much)

CriesofCarrots reviewed (commented):
Just a couple small comments on the cli integration. I'll defer to pgarg66 and lijunwangs as to whether the actual StakedNodes handling is correct.

KirillLykov commented:
@lijunwangs I've got rid of total_stake and stake cli args and, instead, added code which gets this information from the network.

lijunwangs commented on bench-tps/src/main.rs:
```diff
@@ -19,43 +18,118 @@ use {
+fn find_node_activated_stake(
+    rpc_client: Arc<RpcClient>,
+    node_id: Pubkey,
+) -> Result<(u64, u64), ()> {
```
Please document the what are the return values

KirillLykov replied:
done

lijunwangs commented on bench-tps/src/main.rs:
```diff
@@ -19,43 +18,118 @@ use {
+        .current
+        .iter()
+        .find(|&vote_account| vote_account.node_pubkey == node_id_as_str)
+        .map_or(Err(()), |value| {
```
Can we log this error if node is not found for better debug?

KirillLykov replied:
done

lijunwangs commented on bench-tps/src/main.rs:
```diff
@@ -19,43 +18,118 @@ use {
+        return ConnectionCache::with_udp(tpu_connection_pool_size);
+    }
+    if client_node_id.is_none() {
+        return ConnectionCache::new_with_client_options(
```
probably neater to call the ::new function like before in this case.

KirillLykov replied:
done

lijunwangs commented on bench-tps/src/cli.rs:
```diff
@@ -513,5 +534,55 @@ pub fn extract_args(matches: &ArgMatches) -> Config {
 
+    if let Some(addr) = matches.value_of("bind_address") {
+        args.bind_address = solana_net_utils::parse_host(addr).unwrap_or_else(|e| {
+            eprintln!("failed to parse bind_address address: {e}");
```
--> Failed to parse bind_address:

KirillLykov replied:
done

lijunwangs commented on bench-tps/src/cli.rs:
```diff
@@ -513,5 +534,55 @@ pub fn extract_args(matches: &ArgMatches) -> Config {
+    if let Ok(node_id) = read_keypair_file(node_id_path) {
+        args.client_node_id = Some(node_id);
+    } else if matches.is_present("client_node_id") {
+        panic!("could not parse identity path");
```
-> Could not parse identity path

Can you use match and actually print the real error returned from read_keypair_file as there can be multiple reasons for the errors like permission, file not found , not protected and etc.?

KirillLykov replied:
I think I can use clap validator for this case. I want to rewrite this cli parsing code in the follow up PR completely.

lijunwangs commented on bench-tps/src/cli.rs:
```diff
@@ -353,6 +357,23 @@ pub fn build_args<'a>(version: &'_ str) -> App<'a, '_> {
+                .value_name("PATH")
+                .takes_value(true)
+                .requires("json_rpc_url")
+                .help("File containing a node id (keypair) of a validator with active stake. This allows communicating with network using staked connection"),
```
-> "File containing the node identity keypair ..."

KirillLykov replied:
done

lijunwangs commented on bench-tps/src/main.rs:
```diff
@@ -19,43 +18,118 @@ use {
+    let total_active_stake: u64 = vote_accounts
+        .current
+        .iter()
+        .chain(vote_accounts.delinquent.iter())
```
I am not familiar with this business logic, why do we add the delinquent here?

KirillLykov replied:
I thought total active stake includes delinquent nodes stake (by doing experiments primarily). And also here it is done this way https://github.com/solana-labs/solana/blame/5108350710b42f8bc2950395a067dc1d15213ba4/cli/src/cluster_query.rs#L1903

KirillLykov replied:
Total stake is coming from `StakedNodes` structure https://github.com/solana-labs/solana/blob/master/streamer/src/streamer.rs#L30, @lijunwangs do you think it is supposed to take into account delinquent nodes stake?

lijunwangs replied:
I am not sure, the current validator code does not include the delinquent stake I think. Can you use delinquent stake to do communications/vote? If so, it should be included, otherwise we probably should not include.

KirillLykov replied:
If I understand the way it works in the TPU side, it looks like there is no filter on delinquent https://github.com/solana-labs/solana/blob/master/core/src/staked_nodes_updater_service.rs#L82
But I've asked on [discord](https://discord.com/channels/428295358100013066/977244255212937306/1076092681706557491).

KirillLykov replied:
So it looks like although atm delinquent stakes are taken into account, they are going to be unstaked https://github.com/solana-labs/solana/issues/24302
Due to that, will change this code to take into account only normal nodes

lijunwangs commented on bench-tps/src/cli.rs:
```diff
@@ -353,6 +357,24 @@ pub fn build_args<'a>(version: &'_ str) -> App<'a, '_> {
+                .takes_value(true)
+                .requires("json_rpc_url")
+                .validator(is_keypair)
+                .help("File containing the node id (keypair) of a validator with active stake. This allows communicating with network using staked connection"),
```
id --> identity. Normally id means identification in computer programs.

KirillLykov replied:
done

lijunwangs commented on bench-tps/src/main.rs:
```diff
@@ -19,43 +18,118 @@ use {
+    match find_result {
+        Some(value) => Ok((value.activated_stake, total_active_stake)),
+        None => {
+            error!("failed to find stake for requested node");
```
---> "Failed to find stake for requested node"

KirillLykov replied:
will address in follow up

lijunwangs reviewed (approved):
Looks good other than the id vs identity naming. I look forward to another PR for it.
